File: sunflower-core/src/main/java/org/tdf/sunflower/consensus/pos/PoSValidator.java
package org.tdf.sunflower.consensus.pos;

import lombok.Setter;
import org.tdf.common.util.HexBytes;
import org.tdf.sunflower.consensus.AbstractValidator;
import org.tdf.sunflower.facade.RepositoryReader;
import org.tdf.sunflower.state.Account;
import org.tdf.sunflower.state.StateTrie;
import org.tdf.sunflower.types.*;

@Setter
public class PoSValidator extends AbstractValidator {
    private final PoSMiner posMiner;

    public PoSValidator(StateTrie<HexBytes, Account> accountTrie, PoSMiner posMiner) {
        super(accountTrie);
        this.posMiner = posMiner;
    }

    @Override
    public ValidateResult validate(RepositoryReader rd, Block block, Block dependency) {
        BlockValidateResult res = super.commonValidate(rd, block, dependency);
        if (!res.getSuccess()) return res;

        if (
            !posMiner.getProposer(dependency, block.getCreatedAt())
                .map(x -> x.getAddress().equals(block.getBody().get(0).getTo()))
                .orElse(false)
        ) return ValidateResult.fault("invalid proposer " + block.getBody().get(0).getSender());
        return res;
    }

    @Override
    public ValidateResult validate(RepositoryReader rd, Header dependency, Transaction transaction) {
        return ValidateResult.success();
    }

    @Override
    public int getChainId() {
        throw new UnsupportedOperationException();
    }
}

File: sunflower-core/src/main/java/org/tdf/sunflower/consensus/pos/PoSValidator.kt
package org.tdf.sunflower.consensus.pos

import org.tdf.common.util.HexBytes
import org.tdf.common.util.HexBytes.Companion.fromBytes
import org.tdf.sunflower.consensus.AbstractValidator
import org.tdf.sunflower.consensus.pow.PoW
import org.tdf.sunflower.facade.RepositoryReader
import org.tdf.sunflower.state.Account
import org.tdf.sunflower.state.StateTrie
import org.tdf.sunflower.types.Block
import org.tdf.sunflower.types.ValidateResult
import org.tdf.sunflower.types.ValidateResult.Companion.fault

class PoSValidator(accountTrie: StateTrie<HexBytes, Account>, private val pos: PoS, override val chainId: Int) :
    AbstractValidator(accountTrie) {
    override fun validate(rd: RepositoryReader, block: Block, dependency: Block): ValidateResult {
        val res = super.commonValidate(rd, block, dependency)
        if (!res.success) return res

        val p = pos.getProposer(rd, dependency, block.createdAt)
        val eq = p?.first == block.coinbase

        if (!eq) {
            return fault("invalid proposer " + block.body[0].sender)
        }

        // validate pow
        val diff = pos.getDifficulty(rd, dependency)


        return if (PoW.compare(PoW.getPoWHash(block), diff.byte32) > 0) fault(
            String.format(
                "nbits validate failed hash = %s, nbits = %s",
                fromBytes(PoW.getPoWHash(block)),
                fromBytes(diff.byte32)
            )
        ) else res
    }
}